src/ledger/LedgerManagerImpl.cpp

```diff
@@ -1759,6 +1762,416 @@ LedgerManagerImpl::prefetchTransactionData(AbstractLedgerTxnParent& ltx,
+                    }
+                    else
+                    {
+                        readOnlyTtlExtensions.emplace(
+                            lk, updatedLe->data.ttl().liveUntilLedgerSeq);
+                    }
+                }
+                else
+                {
+                    // A entry deletion will be marked by a nullopt le.
+                    // Set the dirty bit so it'll be written to ltx later.
+                    it->second = {updatedLe, true};
+                }
+            }
+
+            for (auto const& key : res.getRestoredKeys().hotArchive)
+            {
+                auto [_, inserted] = threadRestoredKeys.hotArchive.emplace(key);
+                releaseAssert(inserted);
+            }
+            for (auto const& key : res.getRestoredKeys().liveBucketList)
+            {
+                auto [_, inserted] =
+                    threadRestoredKeys.liveBucketList.emplace(key);
+                releaseAssert(inserted);
+            }
+        }
+        else
+        {
+            releaseAssert(!txBundle.getResPayload()->isSuccess());
+        }
+    }
+
+    for (auto const& kvp : readOnlyTtlExtensions)
+    {
+        auto const& lk = kvp.first;
+        auto const& ttlExtension = kvp.second;
+
+        auto it = entryMap.find(lk);
+        // The entry in entryMap should always exist. If the entry was deleted,
+        // then we would've erased the TTL key from readOnlyTtlExtensions.
+        releaseAssert(it != entryMap.end() && it->second.mLedgerEntry);
+        if (it->second.mLedgerEntry->data.ttl().liveUntilLedgerSeq <
+            ttlExtension)
+        {
+            it->second.mLedgerEntry->data.ttl().liveUntilLedgerSeq =
+                ttlExtension;
+            it->second.isDirty = true;
+        }
+    }
+
+    return threadRestoredKeys;
+}
+
+ParallelLedgerInfo
+getParallelLedgerInfo(AppConnector& app, AbstractLedgerTxn& ltx)
+{
+    auto const& lh = ltx.loadHeader().current();
+    return {lh.ledgerVersion, lh.ledgerSeq, lh.baseReserve,
+            lh.scpValue.closeTime, app.getNetworkID()};
+}
+
+void
+LedgerManagerImpl::applySorobanStage(AppConnector& app, AbstractLedgerTxn& ltx,
+                                     ApplyStage const& stage,
+                                     Hash const& sorobanBasePrngSeed)
+{
+    for (auto const& thread : stage)
+    {
+        for (auto const& txBundle : thread)
+        {
+            // This can mark a tx as failed due to validation.
+            // Make sure results are checked.
+            txBundle.getTx()->preParallelApply(
+                app, ltx, txBundle.getEffects().getMeta(),
+                txBundle.getResPayload(),
+                txBundle.getEffects().getTxEventManager());
+        }
+    }
+
+    auto const& config = app.getConfig();
+    auto const& sorobanConfig = app.getSorobanNetworkConfigForApply();
+    auto& sorobanMetrics = app.getLedgerManager().getSorobanMetrics();
```

Comment:
Looks like we can get here not on the main thread with background ledger close, which will lead to a failed asset in AppConnector.